[Dynamis Helper] Fix, Enhance & Refactor

loicteixeira opened the pull request:
4c18a13 has all the original work and requested changed from #1519.
a7794d4 splits the data from the code (as a separate commit so it can be removed if you don't like the idea).

![The new interface with 4 vertical sections: time remaining, trigger type, currency drops and time extensions](https://user-images.githubusercontent.com/6834938/98438813-02213200-20ed-11eb-9d13-14d228f20a64.png)

FWIW, I observed one issue, when you target a monster with no trigger (or a player) it will show the trigger type of the previous monster. But tbh, the infobar addon has the same issue and the dynamishelper addon currently on master is non-functionnal so the new version is already a big improvement.

loicteixeira committed:
Refactor & Add Nicer Interface

Original work Windower/Lua#1519

- Add nicer interface with more options
- Compatible with dynamis campaign
event
- Stagger timer now displays the mobs name and the stun icon
- Stagger timer now displays the mobs name and the stun icon
- New commands are now compatible to the original version (old settings are converted)
- Time extensions are now reseted on zone change
- General refactoring of code and optimization

Co-Authored-By: Sascha Tausend <sascha.tausend@gmail.com>

loicteixeira committed:
Split data & code

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+_addon.author = 'Krizz, Skyrant'
 _addon.commands = {'DynamisHelper','dh'}
-_addon.version = '1.0.2.0'
+_addon.version = '2.3'
```
Feels like a bit of an arbitrary jump, not that I mind :)

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
 --(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 --SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 --Features
```
This is not your fault as it has been there before, but add a spacer line after the copyright disclaimer. That took me a while to mentally parse.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
-							 	"Nightmare Hornet", "Nightmare Bugard", 
-								"Shamblix Rottenheart", "Elvaansticker Bxafraff", "Qu'Pho Bloodspiller", "Te'Zha Ironclad", "Koo Rahi the Levinblade", 
-								"Barong", "Alklha", "Stihi", "Fairy Ring", "Stcemqestcint", "Stringes", "Suttung" }
+ProcZones = res.zones:english(string.startswith-{'Dynamis'}):keyset()
```
You may be excused for thinking we don't care much for naming seeing how the entire repository is littered with inconsistencies, but we have been trying to enforce it for a while now. For settings names you're free to use *PascalCase*, otherwise please stick to *snake_case*. This was not your doing as it has existed before, but if you're in the process of refactoring it I would like to ask you to adjust this as well, but it's up to you.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
-staggers['day']['none'] = {"Animated Claymore", "Animated Dagger", "Animated Great Axe", "Animated Gun", "Animated Hammer", "Animated Horn", "Animated Kunai", "Animated Knuckles", "Animated Longbow", "Animated Longsword", "Animated Scythe", "Animated Shield", "Animated Spear", "Animated Staff", "Animated Tabar", "Animated Tachi", "Fire Pukis", "Petro Pukis", "Poison Pukis", "Wind Pukis", "Kindred ...
+-------------------------------------------------------------------------------
+-- Define default values ------------------------------------------------------
+-------------------------------------------------------------------------------
```
Also not a huge issue, but one line comments are already too much and are only wasting space, three lines is just showing off :) Goes for the other comments, too. Remove the header/footer comments, and preferably also the content comment, though I will leave that one up to you.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+    settings.window.pos.x = settings.trposx
+    settings.window.pos.y = settings.trposy
+    -- delete old settings here -----------------------------------------------
+    config.save(settings, 'all')
```
This is quite dangerous and a deal breaker for this PR until it is adjusted. It would overwrite all user settings for all characters. We don't have a mechanism to update settings for all characters, so either you have to read the XML manually and rewrite it before load, or you just update one character at a time, leaving the XML in an inconsistent state. While it sounds bad it shouldn't be an issue, because each character section should be overwritten once the addon is loaded for that character.

I'm just not 100% sure if existing settings can be removed from the XML, but I think it should be possible by just setting them to `nil`. Even if not, it's not a massive deal, because they will just be ignored in the addon.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
-								"Nightmare Leech", "Nightmare Worm", 
-							 	"Woodnix Shrillwhistle", "Hamfist Gukhbuk", "Lyncean Juwgneg", "Va'Rhu Bodysnatcher", "Doo Peku the Fleetfoot",
-							 	"Nant'ina", "Antaeus"}
+Green = "\\cs(0,255,0)"
```
Please use single quotes for all strings where it makes sense. We allow exceptions for log output or item names, where apostrophes may be regularly encountered, otherwise we try to be consistent here. Also, please `snake_case` the variable name.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+-- Initialize the Currencies array. We need this to keep track of the drops -----
+-------------------------------------------------------------------------------
+function init_currency()
+	for currency in Currencies:it() do
```
A tab managed to sneak in.

z16 commented on addons/dynamishelper/statics.lua:
```diff
@@ -0,0 +1,1005 @@
+-- Currencies -----------------------------------------------------------------
+-------------------------------------------------------------------------------
+
+Currencies = T {
```
You're using `T`-tables for everything here, but I don't know if many of them even need it. Those are always associated with a performance cost. It is a pretty minor issue since the performance cost is *mostly* on the initialization and these here are only initialized once. `Currencies` is iterated with `:it()` above, so it actually makes use of it, this was just meant as a comment on the file in general.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+-------------------------------------------------------------------------------
+function init_window()
+    local showCurrenciesDivider = false
+    window.text(window,Yellow.."Time remaining: ${time|initializing...}")
```
While this is technically fine, it's meant to be called like this:
```lua
window:text(Yellow .. 'Time remaining: ${time|initializing...}')
```

If you call it the way you do now you can call the library function directly:
```lua
texts.text(window, Yellow .. 'Time remaining: ${time|initializing...}')
```

This saves one indirection. Making a function taking the associated object as the first parameter is commonly used in Lua to enable method-like syntax with `object:method(params)`.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
-staggers['night']['random'] = {"Nightmare Taurus"}
-staggers['night']['none'] = {"Animated Claymore", "Animated Dagger", "Animated Great Axe", "Animated Gun", "Animated Hammer", "Animated Horn", "Animated Kunai", "Animated Knuckles", "Animated Longbow", "Animated Longsword", "Animated Scythe", "Animated Shield", "Animated Spear", "Animated Staff", "Animated Tabar", "Animated Tachi", "Fire Pukis", "Petro Pukis", "Poison Pukis", "Wind Pukis", "Kindr ...
+window = texts.new(" ",settings.window,settings)
+w = T{}
```
Bit of a confusing variable name. What is this table meant to contain? Window-specific information? Also, seems like the `T` is unnecessary here.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
-proc = settings['proc']
-pposx = settings['pposx']
-pposy = settings['pposy']
+window:hide()
```
This is fine, just feels a bit out of place. I would move this line to where the window is created, to make it obvious that it's meant to be invisible from the start. Although I thought they are invisible by default, but not sure.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+-------------------------------------------------------------------------------
+-- Check if we are in Dynamis and show the overlay ----------------------------
+-------------------------------------------------------------------------------
+if ProcZones:contains(windower.ffxi.get_info().zone) then
```
I feel like this `if` block can be removed and the `zone change` and `load` and `login` events should be merged. Would save on a bit of code duplication and also put the initialization logic all in one place, since they are all used to determine initialization.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+--  39  Dynamis - Valkurm -----------------------------------------------------
+-- 187  Dynamis - Windurst ----------------------------------------------------
+-- 135  Dynamis - Xarcabard ---------------------------------------------------
+-------------------------------------------------------------------------------
```
Not sure what purpose this comment servers? Or just for debugging?

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+            return new, color
+        end
+    end
+    if string.find(original,"remaining in Dynamis.") then
```
There are just a few inconsistencies in method calls. The original addon uses `string.find(original, 'something')` whereas you seem to use `original:match('something')`, but left the old code in the old style, which results in mixed invocation styles within the file. I would recommend to use the latter style for every call, but I'll leave this up to you.

Style aside, I would suggest to match the strings as closely as possible. This would also match if a player wrote "Not much longer remaining in Dynamis. We should hurry.". I would check the original mode (called `color` here) as well as matching the text as tightly as possible. I don't know the exact message, but iirc it goes something like "30 minutes remaining in Dynamis.", which I would parse with `original:endswith("%d+ minutes remaining in Dynamis.")` or something like that. If you need to alternate between `minutes` and `hours` you can use our regex matching library (not sure if the game reports hours as well, the code afterwards suggests it does not, just mentioned it in case I'm not remembering it correctly).

This goes for the next few `string.find` checks as well.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+    end
+    if string.find(original,"Your stay in Dynamis has been extended by %d+ minutes.") then
+        end_time = end_time + (tonumber(original:match("%d+")) * 60)
+        w.time = os.date('!%H:%M:%S', end_time)
```
I notice this case adjusts `w.time`, but the others do not. This one does not adjust `ob_time` though. Is that correct?

loicteixeira replied:
The 2 cases above are statements of how much time left, not time extensions like this one, so there should be no need to update `w.time`. But it would also mean that there is no need to update the internal objects, which leads me to believe it was made in case the plugin is loaded after entering Dynamis.
Anyways, `obj_time`, `end_time` and `w.time` are all linked, so I'll update them together. I'll also rename `obj_time` to `time_remaining_in_seconds` or something alike.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+    end
+    item = original:match("Obtained key item: ..(%w+ %w+ %w+ %w+)..\46")
+    if item ~= nil then
+        item = item:lower()
```
This is also not your code, but there's really no reason to use `:lower()` on both strings. We get them directly from the game, so they should match case as well.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+            --a,b,item = string.find(original,"%w+ obtains an? ..(%w+%-%w+ %w+)..\46")
+            if item == nil then
+                item = original:match("%w+ obtains an? ..(%w+ %w+)..\46")
+                --a,b,item = string.find(original,"%w+ obtains an? ..(%w+ %w+)..\46")
```
While I like tight matching, this may be overdoing it :) If you want to catch all those individual cases you could as well just check for the individual items directly:
```lua
if original:match('%w+ obtains a 100 Byne Bill') or
   original:match('%w+ obtains an Ordelle Bronzepiece') or
   -- and so on
```

And in that case you might as well use your `Currencies` table:
```lua
if Currencies:any(function(currency) return original:match('%w+ obtains an? ..' .. currency .. '..\46') end) then
```

It does not get tighter than that. But will that not trigger for messages you receive for other characters as well? This is also not just your code, if it is an issue it was there before your refactoring as well. Or is that expected behavior?

This is honestly something that would be better solved using packets, but I guess that's outside of the scope of this refactoring.

Tl;dr I would either just match it more loosely as it was originally, or use exact matching using the table.

loicteixeira replied:
Realistically, this plugin will be used for solo farming, so there's no risk of it matching another character in this situation.
However, nothing stops you from using it with a group (or while multi-boxing) and in both cases, tracking currencies just for you or for the group (so you have the total amount to split at the end) are both valid cases.
This could be made configurable, but I'm running out of time and will therefore change it to only get your own drops. Someone else can contribute later if they want it configurable.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+        end
+    end
+    if item ~= nil then
+        item = item:lower()
```
See above about `:lower()`.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+-------------------------------------------------------------------------------
+-- Register target change event to get the monster name -----------------------
+-------------------------------------------------------------------------------
 windower.register_event('target change', function(targ_id)
```
Also not your thing, but this variable name is wrong. It should be `targ_index` (or preferably `target_index`), it's not an ID.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+-------------------------------------------------------------------------------
+-- Print the help Information -------------------------------------------------
+-------------------------------------------------------------------------------
+help = T{
```
This also does not need to be a `T`-table, and the contents should be indented.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+
+function printHelp(command,val)
+    if not command and not val then
+        windower.add_to_chat(159,'\nDynamisHelper v2.0')
```
`2.0` is not quite right. You can just use `'DynamisHelper v' .. tostring(_addon.version)`.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+    if command == "help" then
+        printHelp()
+        return
+    elseif command:lower() == "visible" then
```
`command` is already lowercased. Also, a reminder about quote style :)

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+        end
+    elseif command == "size" then
+        if options[1] and tonumber(options[1]) then
+            window:size(options[1])
```
Wouldn't this need to be `tonumber(options[1])` as well?

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+            printHelp(command,settings.window.text.font)
+        end
+    elseif command == "posx" then
+        if options[1] then
```
I think this needs `tonumber` as well? Same for some of the next ones. Maybe just go over these once more to make sure.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+    elseif command == "size" then
+        if options[1] and tonumber(options[1]) then
+            window:size(options[1])
+            config.save(settings, 'all')
```
Saving for `'all'` should not be done without a good reason. This disables per-character settings. What was the reason for doing this?

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,427 @@
+        else
+            printHelp(command,"")
+        end
+    elseif command == "ll" then
```
I would love to remove this command entirely. LL has been deprecated for six years and the code in here is messy and not particularly useful. If you remove it, I will accept the PR, but I'll leave this up to you.

If you do keep it, no need to get the player name again, it is already set in the `player` variable on load and login.

z16 reviewed (changes requested):
Sorry for the delay. I added a number of comments with a few things that may need to be discussed. Feel free to respond to the points I made here or message me on Discord.

loicteixeira commented:
Unfortunately I cannot answer the why questions as I was merely trying to push the original PR to the finish line by applying the existing comments.

Thank you for your time, I'll make sure it was not wasted and push an update in the coming weeks.

z16 commented:
Hm, that explains a bit. I was meaning to ask why not make it a completely new addon, since it seems to completely rewrite the original. Not that I mind, though I think it would have had a good case for being its own addon.

loicteixeira committed:
Styling & fixes

- Remove unnecessary comments
- Fix version number
- Fix styling (use single quotes & snake case)
- Do not overwrite all settings
- Call methods with `obj:method(*args)` instead of `obj.method(obj, *args)

loicteixeira commented:
Still working through the changes.

I actually got burned out by the game and cancelled my subscription, hence why I didn't come to this sooner. I still have a week left, so I'll try to take care of the latest required changes (and test them) before then.

As for creating a separate addon, the existing one is non-functional, so I reckon we might as well overwrite it instead of creating a new one.

loicteixeira committed:
Style & fixes

- Tabs to spaces
- Always update `obj_time`, `end_time` and `w.time` together
- Rename `obj_time` to `time_remaining_in_seconds` and `w` to `state` for clarity
- Use tighter pattern matching
- Check that arguments for `posx`, `posy` and `padding` commands are indeed numerical

loicteixeira committed:
Fixes and tighter regexp matches

loicteixeira committed:
Fixes

- Merge zone_change, login and load events together
- Only initialize from within the corresponding events, not globally
- Tweak window update so it works after reloading while in Dynamis
- Break early once we've found a granule/currency match

loicteixeira commented:
I addressed all the issues, except for not using T-tables.

FWIW, time tracking does not work when the plugin is loaded after entering dynamis, which makes sense because how would the plugin now how much time is left? Aside from that, it tracks drops and displays stagger trickers properly.

Unfortunately my subscription is expiring tomorrow, so I'm afraid I won't be able to make further changes. Feel free to close the PR then unless someone wants to pick it up like I pick up the previous PR.

Happy gaming everybody.
